### ISSUE TITLE
sphinx build fails with v0.9

### ISSUE THREAD

**Issue by danielvasas** [sphinx-contrib/plantuml#5]
Hi,

after upgrading to sphinxcontrib-plantuml 0.9, I'm getting following exception:

```
# Sphinx version: 1.6.6
# Python version: 3.6.2 (CPython)
# Docutils version: 0.14 
# Jinja2 version: 2.10
# Last messages:
#   writing output... [  6%] contents/doc_dev/applications/VerifikatorApp/api/index
#   
#   writing output... [ 10%] contents/doc_dev/applications/VerifikatorApp/architecture/index
#   
#   writing output... [ 13%] contents/doc_dev/applications/VerifikatorApp/index
#   
#   writing output... [ 16%] contents/doc_dev/applications/VerifikatorApp/traceability/index
#   
#   writing output... [ 20%] contents/doc_dev/applications/VerifikatorApp/traceability/overview/index
#   
# Loaded extensions:
#   alabaster (0.7.10) from c:\dev\ve_safe_build_36\lib\site-packages\alabaster\__init__.py
#   sphinx.ext.autodoc (1.6.6) from c:\dev\ve_safe_build_36\lib\site-packages\sphinx\ext\autodoc.py
#   sphinxcontrib.plantuml (unknown version) from c:\dev\ve_safe_build_36\lib\site-packages\sphinxcontrib\plantuml.py
#   sphinxcontrib.needs (0.1.47) from c:\dev\ve_safe_build_36\lib\site-packages\sphinxcontrib\needs.py
Traceback (most recent call last):
  File "c:\dev\ve_safe_build_36\lib\site-packages\sphinx\cmdline.py", line 306, in main
    app.build(opts.force_all, filenames)
  File "c:\dev\ve_safe_build_36\lib\site-packages\sphinx\application.py", line 339, in build
    self.builder.build_update()
  File "c:\dev\ve_safe_build_36\lib\site-packages\sphinx\builders\__init__.py", line 329, in build_update
    'out of date' % len(to_build))
  File "c:\dev\ve_safe_build_36\lib\site-packages\sphinx\builders\__init__.py", line 395, in build
    self.write(docnames, list(updated_docnames), method)
  File "c:\dev\ve_safe_build_36\lib\site-packages\sphinx\builders\__init__.py", line 432, in write
    self._write_serial(sorted(docnames))
  File "c:\dev\ve_safe_build_36\lib\site-packages\sphinx\builders\__init__.py", line 441, in _write_serial
    self.write_doc(docname, doctree)
  File "c:\dev\ve_safe_build_36\lib\site-packages\sphinx\builders\html.py", line 555, in write_doc
    self.docwriter.write(doctree, destination)
  File "c:\dev\ve_safe_build_36\lib\site-packages\docutils\writers\__init__.py", line 80, in write
    self.translate()
  File "c:\dev\ve_safe_build_36\lib\site-packages\sphinx\writers\html.py", line 56, in translate
    self.document.walkabout(visitor)
  File "c:\dev\ve_safe_build_36\lib\site-packages\docutils\nodes.py", line 174, in walkabout
    if child.walkabout(visitor):
  File "c:\dev\ve_safe_build_36\lib\site-packages\docutils\nodes.py", line 174, in walkabout
    if child.walkabout(visitor):
  File "c:\dev\ve_safe_build_36\lib\site-packages\docutils\nodes.py", line 174, in walkabout
    if child.walkabout(visitor):
  File "c:\dev\ve_safe_build_36\lib\site-packages\docutils\nodes.py", line 166, in walkabout
    visitor.dispatch_visit(self)
  File "c:\dev\ve_safe_build_36\lib\site-packages\docutils\nodes.py", line 1882, in dispatch_visit
    return method(node)
  File "c:\dev\ve_safe_build_36\lib\site-packages\sphinxcontrib\plantuml.py", line 261, in html_visit_plantuml
    for e in fileformats)
  File "c:\dev\ve_safe_build_36\lib\site-packages\sphinxcontrib\plantuml.py", line 261, in <genexpr>
    for e in fileformats)
  File "c:\dev\ve_safe_build_36\lib\site-packages\sphinxcontrib\plantuml.py", line 160, in render_plantuml
    p = subprocess.Popen(generate_plantuml_args(self, node, fileformat),
  File "c:\dev\ve_safe_build_36\lib\site-packages\sphinxcontrib\plantuml.py", line 147, in generate_plantuml_args
    args.extend(['-filename', node['filename']])
  File "c:\dev\ve_safe_build_36\lib\site-packages\docutils\nodes.py", line 567, in __getitem__
    return self.attributes[key]
KeyError: 'filename'
```

With v0.8.1 there is no such problem.

BR,
Daniel

**Comment by yuja** [comment]
I have no idea why `node['filename']` is missing. It's added at fe41f989, but
basically `node['filename']` should be available as long as `node['incdir']` is
there.

Can you make a minimal repro?

**Comment by danielvasas** [comment]
Sure, here you go
[plantuml-repro.zip](https://github.com/sphinx-contrib/plantuml/files/1638322/plantuml-repro.zip)

To reproduce:
1. pip install -r doc-requirements.txt
2. cd docs
3. sphinx-build -b html -W -E . _build/html

**Comment by danwos** [comment]
Can confirm this, same problem here.
Adding the following to our sphinx projects made our builds runnable again:
`sphinxcontrib-plantuml < 0.9`
(This will normally use 0.8.2)

**Comment by danwos** [comment]
Ok, maybe sphinxcontrib-plantuml is not the guilty lib, but [sphinx-needs](http://sphinxcontrib-needs.readthedocs.io/en/latest/).
Luckily I'm the maintainer of this lib and have a guess, whats going wrong.

sphinx-needs creates a plantuml node by itself to show all need-connections in a nice overview picture.
As @yuja  wrote, the "filename" attribute was introduced some weeks ago, so I guess sphinx-needs needs to set this as well.
I think the following code of sphinx-needs needs to be fixed:
https://github.com/useblocks/sphinxcontrib-needs/blob/master/sphinxcontrib/needs/needfilter.py#L132
```
puml_node = plantuml(plantuml_block_text, **dict())
puml_node["uml"] = "@startuml\n"
puml_node["filename"] = ""  # needed change (untested yet)
```

**Comment by yuja** [comment]
> What do you think, is setting node["filename"] to "" or None enough?

"" should work. Don't set to None as it's passed as a command argument.

> Any other parameters which were add in the past and should be set?

'incdir' was added at 0.8. It can be an empty string, too.

**Comment by danwos** [comment]
Ok, fixed it with [sphinx-needs 0.1.49](https://pypi.python.org/pypi/sphinxcontrib-needs/0.1.49)
Used code is now:
```
puml_node["incdir"] = os.path.dirname(current_needlist["docname"])  # Was working all the time
puml_node["filename"] = os.path.split(current_needlist["docname"])[1]  # Needed for plantuml >= 0.9
```
It's the same logic as used by plantuml itself.
See related commit https://github.com/sphinx-contrib/plantuml/commit/fe41f989c00b64e416a719481af8994f67d3dc2d#diff-c26d945e340e0721a7dc17b686ca702eR94

Sorry for all the confusion and thanks for all the help :+1:

**Comment by yuja** [comment]
Awesome, thanks.
